### PR TITLE
Simplify relationships between processes, templates and workpieces

### PR DESCRIPTION
**Pull request by Beacze**

### CODE REVIEW

File: Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java [hunk 1]
```diff
@@ -2155,7 +2109,7 @@ public void setProperty(Property property) {
     }
 
     /**
-     * Get list of process properties.
+     * Get list of process' properties.
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java]: Is adding of this apostrophe `'` really wanted?

File: Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java [hunk 2]
```diff
@@ -2164,7 +2118,7 @@ public void setProperty(Property property) {
     }
 
     /**
-     * Set list of process properties.
+     * Set list of process' properties.
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java]: Is adding of this apostrophe `'` really wanted?

File: Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java [hunk 3]
```diff
@@ -2173,11 +2127,59 @@ public void setProperties(List<Property> properties) {
         this.properties = properties;
     }
 
+    /**
+     * Get list of process' templates.
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java]: Is adding of this apostrophe `'` really wanted?

File: Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java [hunk 4]
```diff
@@ -2173,11 +2127,59 @@ public void setProperties(List<Property> properties) {
...
+     *
+     * @return list of process templates
+     */
+    public List<Property> getTemplates() {
+        return this.templates;
+    }
+
+    /**
+     * Set list of process' templates.
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java]: Is adding of this apostrophe `'` really wanted?

File: Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java (hunk 1 above)
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java]: Is adding of this apostrophe `'` really wanted?

File: Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java [hunk 5]
```diff
@@ -2173,11 +2127,59 @@ public void setProperties(List<Property> properties) {
...
+     *
+     * @param templates
+     *            for process as Property objects
+     */
+    public void setTemplates(List<Property> templates) {
+        this.templates = templates;
+    }
+
+    /**
+     * Get list of process' workpieces.
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java]: Is adding of this apostrophe `'` really wanted?

File: Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java [hunk 6]
```diff
@@ -2173,11 +2127,59 @@ public void setProperties(List<Property> properties) {
...
+     *
+     * @return list of process' workpieces
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java]: Is adding of this apostrophe `'` really wanted?

File: Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java [hunk 7]
```diff
@@ -2173,11 +2127,59 @@ public void setProperties(List<Property> properties) {
...
+     */
+    public List<Property> getWorkpieces() {
+        return this.workpieces;
+    }
+
+    /**
+     * Set list of process' workpieces.
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java]: Is adding of this apostrophe `'` really wanted?

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java [hunk 8]
```diff
@@ -470,6 +454,18 @@ public QueryBuilder getQueryProjectTitle(String title) {
         return searcher.findDocuments(query.toString());
     }
 
+    public List<JSONObject> findByProcessProperty(String title, String value, boolean contains) throws DataException {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]: New public method without JavaDoc - please add JavaDoc.

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java [hunk 9]
```diff
@@ -470,6 +454,18 @@ public QueryBuilder getQueryProjectTitle(String title) {
...
+        return findByProperty(title, value, "process", "properties.id", contains);
+    }
+
+    public List<JSONObject> findByTemplateProperty(String title, String value, boolean contains) throws DataException {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]: New public method without JavaDoc - please add JavaDoc.

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java [hunk 10]
```diff
@@ -470,6 +454,18 @@ public QueryBuilder getQueryProjectTitle(String title) {
...
+        return findByProperty(title, value, "template", "templates.id", contains);
+    }
+
+    public List<JSONObject> findByWorkpieceProperty(String title, String value, boolean contains) throws DataException {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]: New public method without JavaDoc - please add JavaDoc.

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java [hunk 11]
```diff
@@ -481,22 +477,22 @@ public QueryBuilder getQueryProjectTitle(String title) {
      *            true or false
      * @return list of JSON objects with processes for specific property
      */
-    List<JSONObject> findByProperty(String title, String value, boolean contains) throws DataException {
+    private List<JSONObject> findByProperty(String title, String value, String type, String key, boolean contains) throws DataException {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]: You are changing method signature without updating the JavaDoc comment - please update the JavaDoc comment.

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java [hunk 12]
```diff
@@ -1710,6 +1706,14 @@ public DigitalDocument getDigitalDocument(Process process) throws PreferencesExc
         return new ArrayList<>(filteredList);
     }
 
+    /**
+     * Count amount of process' templates.
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]: Is adding of this apostrophe `'` really wanted?

File: Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java [hunk 13]
```diff
@@ -1710,6 +1706,14 @@ public DigitalDocument getDigitalDocument(Process process) throws PreferencesExc
...
+     * @return amount of process' templates as Long
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/services/data/ProcessService.java]: Is adding of this apostrophe `'` really wanted?

File: Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Vorlagen.xhtml [hunk 14]
```diff
@@ -11,70 +11,166 @@
  *
 -->
 
-<!-- ===================== GROSSE Vorlagen- BOX IM GELADENEN PROZESS ====================== -->
+<!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
+<!-- +++++++++++     Eigenschaftentabelle      ++++++++++ -->
```
Comment (henning-gerhardt) [inline Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Vorlagen.xhtml]: Please use English comments.

File: Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Vorlagen.xhtml [hunk 15]
```diff
@@ -11,70 +11,166 @@
...
-                   value="#{msgs.vorlageHinzufuegen}" title="#{msgs.vorlageHinzufuegen}">
+    <h:commandLink action="#{ProzessverwaltungForm.createTemplateProperty}" value="#{msgs.eigenschaftHinzufuegen}"
+                   title="#{msgs.eigenschaftHinzufuegen}"
+                   rendered="#{ProzessverwaltungForm.editMode.mode != 'templateProperty'}">
+        <t:updateActionListener property="#{ProzessverwaltungForm.editMode}" value="#{objectMode.TEMPLATEPROPERTY}"/>
     </h:commandLink>
+
+
+    <!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
+    <!-- +++++++++     Eigenschaft bearbeiten      ++++++++++ -->
```
Comment (henning-gerhardt) [inline Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Vorlagen.xhtml]: Please use English comments.

File: Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Vorlagen.xhtml [hunk 16]
```diff
@@ -11,70 +11,166 @@
...
+    <!-- Box für die Bearbeitung der Details -->
+    <ui:fragment rendered="#{ProzessverwaltungForm.editMode.mode == 'templateProperty'}">
+        <table cellpadding="3" cellspacing="0" width="100%" class="eingabeBoxen">
+            <tr>
+                <td class="eingabeBoxen_row1" colspan="2">
+                    <h:outputText value="#{msgs.eigenschaft}"/>
+                </td>
+            </tr>
+
+            <!-- Formular für die Bearbeitung der Eigenschaft -->
```
Comment (henning-gerhardt) [inline Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Vorlagen.xhtml]: Please use English comments.

File: Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Vorlagen.xhtml [hunk 17]
```diff
@@ -11,70 +11,166 @@
...
+    <!-- +++++++++     Eigenschaft bearbeiten      ++++++++++ -->
+    <!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
+
+    <ui:fragment rendered="#{ProzessverwaltungForm.editMode.mode == 'templateProperty'}">
+        <h4 style="margin-top:15px">
+            <h:outputText value="#{msgs.eigenschaftBearbeiten}"/>
+        </h4>
+    </ui:fragment>
+
+    <!-- Box für die Bearbeitung der Details -->
```
Comment (henning-gerhardt) [inline Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Vorlagen.xhtml]: Please use English comments.

File: Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Werkstuecke.xhtml [hunk 18]
```diff
@@ -11,72 +11,166 @@
  *
 -->
 
-<!-- ===================== GROSSE Werkstuecke- BOX IM GELADENEN PROZESS ====================== -->
+<!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
+<!-- +++++++++++     Eigenschaftentabelle      ++++++++++ -->
```
Comment (henning-gerhardt) [inline Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Werkstuecke.xhtml]: Please use English comments.

File: Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Werkstuecke.xhtml [hunk 19]
```diff
@@ -11,72 +11,166 @@
...
-                   title="#{msgs.werkstueckHinzufuegen}">
+    <h:commandLink action="#{ProzessverwaltungForm.createWorkpieceProperty}" value="#{msgs.eigenschaftHinzufuegen}"
+                   title="#{msgs.eigenschaftHinzufuegen}"
+                   rendered="#{ProzessverwaltungForm.editMode.mode != 'workpieceProperty'}">
+        <t:updateActionListener property="#{ProzessverwaltungForm.editMode}" value="#{objectMode.WORKPIECEPROPERTY}"/>
     </h:commandLink>
+
+
+    <!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
+    <!-- +++++++++     Eigenschaft bearbeiten      ++++++++++ -->
```
Comment (henning-gerhardt) [inline Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Werkstuecke.xhtml]: Please use English comments.

File: Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Werkstuecke.xhtml [hunk 20]
```diff
@@ -11,72 +11,166 @@
...
+    <!-- ++++++++++++++++++++++++++++++++++++++++++++++++++++ -->
+
+    <ui:fragment rendered="#{ProzessverwaltungForm.editMode.mode == 'workpieceProperty'}">
+        <h4 style="margin-top:15px">
+            <h:outputText value="#{msgs.eigenschaftBearbeiten}"/>
+        </h4>
+    </ui:fragment>
+
+    <!-- Box für die Bearbeitung der Details -->
```
Comment (henning-gerhardt) [inline Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Werkstuecke.xhtml]: Please use English comments.

File: Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Werkstuecke.xhtml [hunk 21]
```diff
@@ -11,72 +11,166 @@
...
+    <ui:fragment rendered="#{ProzessverwaltungForm.editMode.mode == 'workpieceProperty'}">
+        <table cellpadding="3" cellspacing="0" width="100%" class="eingabeBoxen">
+            <tr>
+                <td class="eingabeBoxen_row1" colspan="2">
+                    <h:outputText value="#{msgs.eigenschaft}"/>
+                </td>
+            </tr>
+
+            <!-- Formular für die Bearbeitung der Eigenschaft -->
```
Comment (henning-gerhardt) [inline Kitodo/src/main/webapp/pages/inc_Prozessverwaltung/prozess_box_Werkstuecke.xhtml]: Please use English comments.

File: Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java (hunk 2 above)
Comment (Beacze) [reply Kitodo/src/main/java/de/sub/goobi/forms/ProzessverwaltungForm.java]: According to English grammar? Yes ;)

**Pull request merged**